### ISSUE TITLE
ESP32 crashed on BLE scan

### ISSUE THREAD

**Issue by Avenitos** [nkolban/esp32-snippets#763]
Hello!

ESP32 crashed on BLE scan with WiFi connected every 10-50 minuts. :(

> E (218187) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
> E (218187) task_wdt:  - IDLE0 (CPU 0)
> E (218187) task_wdt: Tasks currently running:
> E (218187) task_wdt: CPU 0: wifi
> E (218187) task_wdt: CPU 1: IDLE1
> E (223187) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
> E (223187) task_wdt:  - IDLE0 (CPU 0)
> E (223187) task_wdt: Tasks currently running:
> E (223187) task_wdt: CPU 0: wifi
> E (223187) task_wdt: CPU 1: IDLE1
> ASSERT_PARAM(-218959118 0), in arch_main.c at line 322
> Guru Meditation Error: Core  0 panic'ed (Interrupt wdt timeout on CPU0)
> Core 0 register dump:
> PC      : 0x40085f4a  PS      : 0x00060834  A0      : 0x801b8350  A1      : 0x3ffbe250  
> A2      : 0x00000001  A3      : 0x00000000  A4      : 0x00000000  A5      : 0x60008054  
> A6      : 0x3ffbdd50  A7      : 0x60008050  A8      : 0x80085f45  A9      : 0x3ffbe230  
> A10     : 0x00000004  A11     : 0x00000000  A12     : 0x6000804c  A13     : 0xffffffff  
> A14     : 0x00000000  A15     : 0xfffffffc  SAR     : 0x00000004  EXCCAUSE: 0x00000005  
> EXCVADDR: 0x00000000  LBEG    : 0x40085e7d  LEND    : 0x40085e84  LCOUNT  : 0x00000000  
> Core 0 was running in ISR context:
> EPC1    : 0x40009203  EPC2    : 0x00000000  EPC3    : 0x00000000  EPC4    : 0x40085f4a
> 
> Backtrace: 0x40085f4a:0x3ffbe250 0x401b834d:0x3ffbe270 0x40019fb5:0x3ffbe290 0x40046683:0x3ffbe2c0 0x40047515:0x3ffbe2e0 0x40087c99:0x3ffbe300 0x400887c2:0x3ffbe320 0x400896fe:0x3ffbe340 0x4008a837:0x3ffbe360 0x4008177d:0x3ffbe380 0x4008b203:0x00000000
> 
> Core 1 register dump:
> PC      : 0x401523e2  PS      : 0x00060e34  A0      : 0x800929d0  A1      : 0x3ffbc610  
> A2      : 0x00000008  A3      : 0x00000001  A4      : 0x00000001  A5      : 0x80000020  
> A6      : 0x00000000  A7      : 0x3ffbd60c  A8      : 0x3ffc9c08  A9      : 0x3ffc9bcc  
> A10     : 0x00000000  A11     : 0x00000001  A12     : 0x80090cc0  A13     : 0x3ffd1c00  
> A14     : 0x00000000  A15     : 0x3ffbc2e0  SAR     : 0x00000000  EXCCAUSE: 0x00000005  
> EXCVADDR: 0x00000000  LBEG    : 0x00000000  LEND    : 0x00000000  LCOUNT  : 0x00000000  
> 
> Backtrace: 0x401523e2:0x3ffbc610 0x400929cd:0x3ffbc630 0x40090aad:0x3ffbc650
> 
> Rebooting...
> ets Jun  8 2016 00:22:57
> 
> rst:0xc (SW_CPU_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
> configsip: 0, SPIWP:0xee
> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
> mode:DIO, clock div:1
> load:0x3fff0018,len:4
> load:0x3fff001c,len:1324
> load:0x40078000,len:7788
> ho 0 tail 12 room 4
> load:0x40080400,len:6448
> entry 0x400806e8

my simple code:
```
#include <WiFi.h>
#include <BLEDevice.h>
#include <BLEUtils.h>
#include <BLEScan.h>
#include <BLEAdvertisedDevice.h>

int scanTime = 2; //In seconds
BLEScan* pBLEScan;

class MyAdvertisedDeviceCallbacks: public BLEAdvertisedDeviceCallbacks {
    void onResult(BLEAdvertisedDevice advertisedDevice) {
      Serial.printf("Advertised Device: %s \n", advertisedDevice.toString().c_str());
    }
};

void WiFi_Setup() {
  WiFi.begin ("ssid", "password");

  // Wait for connection
  while ( WiFi.status() != WL_CONNECTED ) {
    delay ( 500 );
    Serial.print ( "." );
  }

  Serial.print ( "IP address: " );
  Serial.println ( WiFi.localIP() ); 
}

void setup() {
  Serial.begin(115200);

  BLEDevice::init("");

  WiFi_Setup();
}

void loop() {
  Serial.printf("------------------------- Uptime: %d,     RAM Free: %d\n", millis(), system_get_free_heap_size());
  
  pBLEScan = BLEDevice::getScan(); //create new scan
  //pBLEScan->setAdvertisedDeviceCallbacks(new MyAdvertisedDeviceCallbacks());
  pBLEScan->setActiveScan(true); //active scan uses more power, but get results faster  
  BLEScanResults foundDevices = pBLEScan->start(scanTime);
  Serial.print("Devices found: ");
  Serial.println(foundDevices.getCount());
  Serial.println("Scan done!"); 

  delay(2000);
}
```

**Comment by chegewara** [comment]
`task_wdt: Task watchdog got triggered`
```
E (223187) task_wdt: - IDLE0 (CPU 0)
E (223187) task_wdt: Tasks currently running:
E (223187) task_wdt: CPU 0: wifi
E (223187) task_wdt: CPU 1: IDLE1
```
From this you can see that wifi is causing watchdog triggers.

**Comment by Avenitos** [comment]
Yes, I see that wifi. But this happens only when scanning BLE devices, what can be done? I do not even send any data, just connected to WiFi. Scanning takes from several minutes to an hour, and then hangs or reboots at best: (

**Comment by chegewara** [comment]
Will test for few more hours but so far its working to me:
```
------------------------- Uptime: 3706994,     RAM Free: 77716
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
15:50:41.706 -> Devices found: 10
15:50:41.706 -> Scan done!
------------------------- Uptime: 3711000,     RAM Free: 77676
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
15:50:45.741 -> Devices found: 10
15:50:45.741 -> Scan done!
------------------------- Uptime: 3715006,     RAM Free: 77732
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
15:50:49.723 -> Devices found: 10
15:50:49.723 -> Scan done!
------------------------- Uptime: 3719012,     RAM Free: 77676
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
15:50:53.722 -> Devices found: 10
15:50:53.722 -> Scan done!
------------------------- Uptime: 3723018,     RAM Free: 77716
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
15:50:57.740 -> Devices found: 10
15:50:57.740 -> Scan done!
```

**Comment by chegewara** [comment]
Still nothing, sorry:
```
------------------------- Uptime: 10397333,     RAM Free: 68576
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
17:42:12.031 -> Devices found: 12
17:42:12.031 -> Scan done!
------------------------- Uptime: 10401342,     RAM Free: 68500
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
17:42:16.048 -> Devices found: 12
17:42:16.048 -> Scan done!
------------------------- Uptime: 10405348,     RAM Free: 68408
[W][BLEScan.cpp:78] handleGAPEvent(): ESP_GAP_SEARCH_INQ_CMPL_EVT
17:42:20.067 -> Devices found: 12
17:42:20.067 -> Scan done!
```

**Comment by Avenitos** [comment]
chegewara, WiFi is connected ?
I worked for a maximum of 6 hours, then reboots :(
How to enable debugging like yours?
Tried other cables and other power supplies, the problem is not that. Enabled the display of WiFi events, before the fall, no events (disconnections from the access point for example) occur.

[https://pastebin.com/2LK81qmu](crash 1)
[https://pastebin.com/PEkTGy2P](crash 2)
[https://pastebin.com/aXvKTARb](crash 3)

Uptime in seconds.

**Comment by chegewara** [comment]
In arduino in serial port monitor you have at bottom left checkbox to see current time in debug log.

In crash 1 log i see you can have memory leak.
Try to use this to decode backtrace:
https://github.com/nkolban/esp32-snippets/blob/master/cpp_utils/ArduinoBLE.md#decoding-an-exception-stack-trace

**Comment by Avenitos** [comment]
it is not always possible to detect the moment of the fall, and sometimes it fixates on task_wdt and the module does not reboot with the backtrace output.

Here's a crash in 4 hours with looping:
[crash 4](https://pastebin.com/4FMnJ9pP)
